lib/Dialect/FIRRTL/FIRRTLAnnotations.cpp

```diff
@@ -84,6 +89,21 @@ bool AnnotationSet::applyToOperation(Operation *op) const {
   return op->getAttrDictionary() != before;
 }
 
+bool AnnotationSet::applyToPort(Operation *op, size_t portNo) const {
+  auto before = op->getAttrDictionary();
+  auto *context = op->getContext();
+  auto oldAttrs = op->getAttrOfType<ArrayAttr>(getPortAnnotationAttrName());
+  SmallVector<Attribute> portAnnotations;
+  if (!oldAttrs || oldAttrs.empty())
+    portAnnotations.assign(getPortCount(op), ArrayAttr::get(context, {}));
+  else
+    portAnnotations.append(oldAttrs.begin(), oldAttrs.end());
+  portAnnotations[portNo] = getArrayAttr();
+  auto after = ArrayAttr::get(context, portAnnotations);
+  op->setAttr(getPortAnnotationAttrName(), after);
+  return op->getAttrDictionary() != before;
```

Comment:
Is there a case when this is `false`, unless `getArrayAttr()` is empty ? 
I was thinking this comparison could be expensive and can be avoided ?

Reply:
The check is actually pretty cheap, its just a pointer comparison since the dictionary is interned.  I did switch how this works, instead of comparing the entire dictionary of the op, i just compare the `portAttributes` which we already have in hand:
```c++
auto before = op->getAttrOfType<ArrayAttr>(getPortAnnotationAttrName());
.....
op->setAttr(getPortAnnotationAttrName(), after);
return before != after;
```